File: backend/app/schemas/group.py
from typing import List, Optional
from pydantic import BaseModel, EmailStr
from datetime import datetime


GROUP_ROLES = ("admin", "member")


class UserInGroup(BaseModel):
    id: int
    name: str
    email: str
    model_config = {"from_attributes": True}


class GroupMemberOut(BaseModel):
    id: int
    group_id: int
    user_id: int
    role: str
    status: str = "accepted"
    user: UserInGroup
    model_config = {"from_attributes": True}


class GroupBase(BaseModel):
    name: str


class GroupCreate(GroupBase):
    pass


class GroupUpdate(BaseModel):
    name: Optional[str] = None


class GroupOut(BaseModel):
    id: int
    name: str
    owner_id: int
    created_at: datetime
    my_role: str
    member_count: int
    trip_count: int
    model_config = {"from_attributes": True}


class GroupDetailOut(BaseModel):
    id: int
    name: str
    owner_id: int
    created_at: datetime
    my_role: str
    model_config = {"from_attributes": True}


class GroupInviteRequest(BaseModel):
    email: EmailStr
    role: str = "member"


class GroupRoleUpdateRequest(BaseModel):
    role: str


class GroupTripSummary(BaseModel):
    id: int
    name: str
    start_date: Optional[datetime] = None
    model_config = {"from_attributes": True}


class GroupInvitationOut(BaseModel):
    """Pending group invitation as seen on the invitee's dashboard."""
    id: int
    group_id: int
    role: str
    group: "GroupSummary"
    inviter: Optional["GroupInviterSummary"] = None
    model_config = {"from_attributes": True}


class GroupSummary(BaseModel):
    id: int
    name: str
    model_config = {"from_attributes": True}


class GroupInviterSummary(BaseModel):
    name: str
    email: str
    model_config = {"from_attributes": True}


GroupInvitationOut.model_rebuild()

File: ios/Roammate/Models/Group.swift
import Foundation

/// Roles on a Group. Backend values: "admin", "member".
enum GroupRole: String, Codable {
    case admin
    case member
}

struct TravelGroup: Codable, Identifiable, Hashable {
    let id: Int
    let name: String
    let ownerId: Int
    let createdAt: Date
    let myRole: String
    let memberCount: Int
    let tripCount: Int

    enum CodingKeys: String, CodingKey {
        case id, name
        case ownerId = "owner_id"
        case createdAt = "created_at"
        case myRole = "my_role"
        case memberCount = "member_count"
        case tripCount = "trip_count"
    }
}

struct GroupDetail: Codable, Identifiable, Hashable {
    let id: Int
    let name: String
    let ownerId: Int
    let createdAt: Date
    let myRole: String

    enum CodingKeys: String, CodingKey {
        case id, name
        case ownerId = "owner_id"
        case createdAt = "created_at"
        case myRole = "my_role"
    }
}

struct GroupCreate: Encodable {
    let name: String
}

struct GroupUpdate: Encodable {
    let name: String?
}

struct GroupInviteRequest: Encodable {
    let email: String
    let role: String
}

struct GroupSummary: Codable, Identifiable, Hashable {
    let id: Int
    let name: String
}

struct GroupInviterSummary: Codable, Hashable {
    let name: String
    let email: String
}

struct GroupInvitation: Codable, Identifiable, Hashable {
    let id: Int
    let groupId: Int
    let role: String
    let group: GroupSummary
    let inviter: GroupInviterSummary?

    enum CodingKeys: String, CodingKey {
        case id, role, group, inviter
        case groupId = "group_id"
    }
}

struct GroupTripSummary: Codable, Identifiable, Hashable {
    let id: Int
    let name: String
    let startDate: Date?

    enum CodingKeys: String, CodingKey {
        case id, name
        case startDate = "start_date"
    }
}

struct GroupMember: Codable, Identifiable, Hashable {
    let id: Int
    let groupId: Int
    let userId: Int
    let role: String
    let status: String
    let user: MemberUser

    enum CodingKeys: String, CodingKey {
        case id, role, status, user
        case groupId = "group_id"
        case userId = "user_id"
    }
}
